File: backup_rollback_replace.py
import shutil
from pathlib import Path
import re

def replace_with_backup(file_path, old_pattern, new_text, config):
    if not check_file_permission(file_path, "write", config):
        raise PermissionError(f"Write access denied for {file_path}")
    file_path = Path(file_path).resolve()
    if not file_path.is_file():
        raise ValueError(f"File not found: {file_path}")
    backup_path = file_path.with_suffix(file_path.suffix + ".bak")
    shutil.copy(file_path, backup_path)
    try:
        with open(file_path, "r", encoding="utf-8") as f:
            content = f.read()
        pattern = re.compile(old_pattern, re.DOTALL)
        if not pattern.search(content):
            raise ValueError(f"Pattern '{old_pattern}' not found in {file_path}")
        new_content = pattern.sub(new_text, content)
        with open(file_path, "w", encoding="utf-8") as f:
            f.write(new_content)
        return f"Replaced pattern in {file_path}, backup at {backup_path}"
    except Exception as e:
        shutil.copy(backup_path, file_path)
        raise ValueError(f"Edit failed, restored backup: {str(e)}")

def replace_rollback(file_path, config):
    if not check_file_permission(file_path, "write", config):
        raise PermissionError(f"Write access denied for {file_path}")
    file_path = Path(file_path).resolve()
    backup_path = file_path.with_suffix(file_path.suffix + ".bak")
    if not backup_path.exists():
        raise ValueError(f"No backup found for {file_path}")
    shutil.copy(backup_path, file_path)
    return f"Restored {file_path} from backup"

File: ts-tools/backup_rollback_replace.ts
import * as fs from 'fs';
import * as path from 'path';
import { checkFilePermission } from './check_file_permission';

export function replaceWithBackup(
  filePath: string,
  oldPattern: string,
  newText: string,
  config: any,
): Promise<string> {
  return new Promise((resolve, reject) => {
    checkFilePermission(filePath, 'write', config)
      .then(() => {
        const resolvedPath = path.resolve(filePath);
        const backupPath = resolvedPath + '.bak';
        fs.copyFile(resolvedPath, backupPath, (err) => {
          if (err) {
            reject(err);
          } else {
            fs.readFile(resolvedPath, 'utf-8', (err, data) => {
              if (err) {
                reject(err);
              } else {
                const regex = new RegExp(oldPattern, 's');
                if (!regex.test(data)) {
                  reject(
                    new Error(
                      `Pattern '${oldPattern}' not found in ${filePath}`,
                    ),
                  );
                } else {
                  const newContent = data.replace(regex, newText);
                  fs.writeFile(resolvedPath, newContent, 'utf-8', (err) => {
                    if (err) {
                      fs.copyFile(backupPath, resolvedPath, () => reject(err));
                    } else {
                      resolve(
                        `Replaced pattern in ${filePath}, backup at ${backupPath}`,
                      );
                    }
                  });
                }
              }
            });
          }
        });
      })
      .catch(reject);
  });
}

export function replaceRollback(
  filePath: string,
  config: any,
): Promise<string> {
  return new Promise((resolve, reject) => {
    checkFilePermission(filePath, 'write', config)
      .then(() => {
        const resolvedPath = path.resolve(filePath);
        const backupPath = resolvedPath + '.bak';
        fs.access(backupPath, fs.constants.F_OK, (err) => {
          if (err) {
            reject(new Error(`No backup found for ${filePath}`));
          } else {
            fs.copyFile(backupPath, resolvedPath, (err) => {
              if (err) {
                reject(err);
              } else {
                resolve(`Restored ${filePath} from backup`);
              }
            });
          }
        });
      })
      .catch(reject);
  });
}
